src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -133,7 +133,11 @@ object ExpandWhens extends Pass {
                     case (WInvalid(), WInvalid()) => WInvalid()
                     case (WInvalid(), fv) => ValidIf(NOT(s.pred), fv, tpe(fv))
                     case (tv, WInvalid()) => ValidIf(s.pred, tv, tpe(tv))
-                    case (tv, fv) => Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                    case (tv, fv) => {
+                      // parameterization might result in the same tv, fv -- get rid of unnecessary mux
+                      if (tv != fv) Mux(s.pred, tv, fv, mux_type_and_widths(tv, fv))
+                      else tv
```

Comment:
I think that or a similar utility would be useful no matter what.

There's no reason we have to wait until the end to do constant propagation. The pass could (and should) be expanded to work on mid firrtl and then could be run following Expand Whens. Better yet, it could be run in some restricted form on high firrtl (the proposed propagation in expand whens would be very difficult to do on high firrtl). However, I believe the crux of the reason why you're getting the bad QOR as in your example above is that Constant Propagation isn't doing its job right. And although I don't think it is causing this particular issue (although it might be), we should consider rerunning ConstProp, CSE, and DCE (perhaps in a loop) since they can create work for the other to do. I believe this is a common compiler pattern.

The reason why I am so hesitant to distribute constant propagation functionality throughout is that it violates separation of concerns. I think that Expand Whens is probably already doing too much work and I would love to reduce it's functionality if possible. Furthermore, Constant Propagation is not required for correctness, and thus should be easy to turn off. Distributing it throughout other passes makes that very difficult.

I totally agree with you about readability and not treating this as one big catch-all (although the one big catch-all should be able to do just that, and it's currently not!)